### PR TITLE
[1.12] Implemented LightingEngine to improve multi light change handling

### PR DESCRIPTION
**Pull request by PhiProven**
1.12 version of #3879

### CODE REVIEW

**Comment by williewillus** [comment]
Does the indexed loop actually change anything? 
An enhanced for-loop over an array compiles to the same code as an indexed loop (no iterator is allocated), and looks cleaner in the code

**Comment by PhiProven** [comment]
The `EnumFacing` was actually not used, only the index. Hence I prefer the indexed loop.
It will be used in #4184, but since the index is needed anyway, I personally think an indexed loop is cleaner than `.ordinal()`.
I also assume that the JIT might like the indexed loop a little better: It may suffice to scalarize the state arrays (ie. neighbor chunks, light, etc.). However, they probably won't be scalarized with `.ordinal()` as the index won't be sufficiently static. Probably, the JIT won't care about this at all, but I still think it's a bit nicer than `.ordinal()`.

**Comment by PhiProven** [comment]
Another Vanilla bug that is fixed by this PR:
If you remove a block from a big (stone) platform, the light won't be spread uniformly, but will be centered around some points in the column (see picture).
This is because Vanilla calculates the correct light values (ie. the blocks in the column below the removed block get a skylight value of 15) for all blocks it encounters during spreading, but confines the spreading radius to 16 blocks around the block that originally started spreading. Our `LightingEngine` doesn't have this limitation.

Vanilla:


This PR:

**Comment by PhiProven** [comment]
We *could*  split this into two PRs to make reviewing easier:
* One part that contains only the `LightingEngine` and carries out updates immediately. This should suffice to solve the Vanilla bugs mentioned above and in #3879. However, this will probably be a little slower than Vanilla in the situation of the bug described above: In case of a heightmap change, the whole column is assigned a new source skylight value at once (the source skylight value is calculated on the fly by looking if the block is below or above the hightmap). The Vanilla lighting engine will then process these changes simultaneously, but with the terrible outcome shown above. Our `LightngEngine` doesn't recalculate the overall light value during spreading (that would be unmecessarily slow), hence it would process the updates one after another and therefore be a little slower. (The current PR doesn't have this problem, since all updates are enqueued for checking before the check is actually carried out)
**Edit**:
We could enable delayed processing just for `Chunk.relightBlock(...)` and process the updates before the method exits. This would solve the performance decrease compared to Vanilla, while delayed processing would be confined to a single method.
* A second part that handles the delayed processing of light updates. That is really what makes this PR so much faster than Vanilla in case of multi block changes. However, it is also the part that is hardest to review, as it contains the most Vanilla interactions (We have to make sure that the updates are carried out whenever needed).

should we do that?

File: src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java [hunk 1]
```diff
@@ -689,6 +689,12 @@ public IThreadListener getWorldThread(INetHandler net)
         return sidedDelegate.getWorldThread(net);
     }
 
+    @Nullable
+    public IThreadListener getMinecraftThread()
```
Comment (kashike) [inline src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java:693]: This should be `getClientThread`, in my opinion. The current name suggests that it is for the thread that Minecraft is running on - be that a client or server.

File: src/main/java/net/minecraftforge/fml/common/IFMLSidedHandler.java [hunk 2]
```diff
@@ -73,6 +75,9 @@
 
     IThreadListener getWorldThread(INetHandler net);
 
+    @Nullable
+    IThreadListener getMinecraftThread();
```
Comment (kashike) [inline src/main/java/net/minecraftforge/fml/common/IFMLSidedHandler.java:79]: This should be `getClientThread`, in my opinion. The current name suggests that it is for the thread that Minecraft is running on - be that a client or server.

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java (hunk 1 above)
Comment (covers1624) [reply src/main/java/net/minecraftforge/fml/common/FMLCommonHandler.java:693]: ~~What? Its clearly used to get the server thread as well.~~ Maybe `getMainThreadListener`? and should be expanded to return the server thread on the server.

Edit: -_- im silly :P

**Comment by williewillus** [comment]
Any chance we could maybe see this PR and its sisters released as a separate mod? 

I understand that's almost too giant to review, get in, and maintain over a long period of time, but some mods and players would really benefit from such optimizations.

**Comment by williewillus** [comment]
Any chance we could maybe see this PR and its sisters released as a separate mod? 

I understand that's almost too giant to review, get in, and maintain over a long period of time, but some mods and players would really benefit from such optimizations.

**Comment by mezz** [comment]
I discussed it with PhiPro95 on discord a little while ago and we agree it would be best as a separate mod.
These changes are very interesting but are not a good fit for Forge.

**Pull request closed without merging**